internal/proxy/task_scheduler.go

```diff
@@ -227,15 +227,19 @@ func (queue *dmTaskQueue) Enqueue(t task) error {
 	// This statsLock has two functions:
 	//	1) Protect member pChanStatisticsInfos
 	//	2) Serialize the timestamp allocation for dml tasks
-	queue.statsLock.Lock()
-	defer queue.statsLock.Unlock()
+
 	//1. preAdd will check whether provided task is valid or addable
 	//and get the current pChannels for this dmTask
-	pChannels, dmt, err := queue.preAddPChanStats(t)
+	dmt := t.(dmlTask)
+	pChannels, err := dmt.getChannels()
 	if err != nil {
+		log.Warn("getChannels failed when Enqueue", zap.Any("tID", t.ID()), zap.Error(err))
 		return err
 	}
+
 	//2. enqueue dml task
+	queue.statsLock.Lock()
```

Comment:
move the lock before pChannels, err := dmt.getChannels() to guard the channel cache``

Reply:
channel cache is only used at `getChannels`, and `getChannels` is only used when task enqueues and dequeues, which will never have a race.